src/Hosting/WindowsServices/src/WindowsServiceLifetime.cs

```diff
@@ -35,6 +36,11 @@ public Task WaitForStartAsync(CancellationToken cancellationToken)
             ApplicationLifetime.ApplicationStopping.Register(() =>
             {
                 Logger.LogInformation("Application is shutting down...");
+                Task.Run(Stop); // Transition the service to StopPending.
```

Comment:
Should not be needed, see StopAsync.

Reply:
It isn't necessary, but it has the nice effect of marking the service as StopPending if the service shuts itself down. For instance, without calling Stop as soon as the application is stopping, this application won't display a StopPending status as it's stopping,

```csharp
class Program
{
    static async Task Main(string[] args)
    {
        using var cts = new CancellationTokenSource(5000);
        using var host = Host.CreateDefaultBuilder(args)
            .UseWindowsService()
            .ConfigureServices(services =>
            {
                services.AddHostedService<MyService>();
                services.Configure<HostOptions>(opt => opt.ShutdownTimeout = TimeSpan.FromSeconds(30));
            })
            .Build();

        await host.RunAsync(cts.Token);
    }
}

public class MyService : IHostedService
{
    public Task StartAsync(CancellationToken cancellationToken)
    {
        return Task.CompletedTask;
    }

    public async Task StopAsync(CancellationToken cancellationToken)
    {
        await Task.Delay(5000);
    }
}
```

This would be easier if ServiceBase had a SetStatus method to arbitrarily change the service's status.